presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewQueryOptimizer.java

```diff
@@ -585,17 +596,136 @@ protected Node visitExpression(Expression node, Void context)
             return super.visitExpression(removeExpressionPrefix(node, removablePrefix), context);
         }
 
+        class NonAssociativeFunctionHandler
+        {
+            // TODO: Move COUNT to this handler
+            // TODO: Add many more functions
+
+            private final Map<QualifiedName, Function<Identifier, Boolean>> validateFunctions = ImmutableMap.of(
+                    AVG, this::validateAvg, APPROX_DISTINCT, this::validateApproxDistinct);
+
+            private final Map<QualifiedName, Function<Identifier, Expression>> rewriteFunctions = ImmutableMap.of(
+                    AVG, this::rewriteAvg, APPROX_DISTINCT, this::rewriteApproxDistinct);
+
+            /**
+             * Rewrites the provided argument if it is non-standard (e.g. it cannot be handled by simply re-applying function to derived columns)
+             */
+            Optional<Expression> rewriteFunction(FunctionCall functionCall)
+            {
+                if (!canRewrite(functionCall)) {
+                    return Optional.empty();
+                }
+
+                return Optional.of(rewrite(functionCall));
+            }
+
+            Expression rewrite(FunctionCall functionCall)
+            {
+                QualifiedName functionName = functionCall.getName();
+                Identifier baseTableColumn = (Identifier) functionCall.getArguments().get(0);
+
+                return rewriteFunctions.get(functionName).apply(baseTableColumn);
+            }
+
+            Expression rewriteAvg(Identifier baseTableColumn)
+            {
+                Optional<Identifier> sumDerived = getDerivedFromFunctionCall(SUM, baseTableColumn);
+                Optional<Identifier> countDerived = getDerivedFromFunctionCall(COUNT, baseTableColumn);
+                Optional<Identifier> avgDerived = getDerivedFromFunctionCall(AVG, baseTableColumn);
+
+                if (sumDerived.isPresent() && countDerived.isPresent()) { //sum(sum_res)/sum(count_res)
+                    FunctionCall sum = new FunctionCall(SUM, ImmutableList.of(sumDerived.get()));
+                    FunctionCall count = new FunctionCall(SUM, ImmutableList.of(countDerived.get()));
+                    return new ArithmeticBinaryExpression(DIVIDE, sum, count);
```

Comment:
This is ok when count = 0, because then sum will be NULL. Comparing to the next if case which needs more complex handling